src/box/txn.c

```diff
@@ -273,6 +273,16 @@ diag_set_txn_sign_detailed(const char *file, unsigned line, int64_t signature)
 	      "at %s:%u", (long long)signature, file, line);
 }
 
+struct txn *
+txn_detach(void)
+{
+	struct txn *txn = in_txn();
+	assert(txn != NULL);
+	trigger_clear(&txn->fiber_on_stop);
```

Comment:
Clearing the` fiber_on_yield 'trigger depends on the `memtx_use_mvcc_engine' option and is performed if this parameter is false from another place

Reply:
I see, but still this looks suspicious. We detach a transaction from a fiber, but leave the transaction subscribed to the fiber's on_yield event. It doesn't make any sense. The fiber can die and get reused for some other purpose, but we'll still run some action when it yields... You should probably clear all triggers here and abort the transaction if it doesn't support yields.

Reply:
By looking at the transaction you can't normally say if it supports yields. You would need an ugly hack such as looking at the mvcc option value right here, in detach. You can try to invoke the on_yield triggers right here like if a yield really happened, but it also looks like a crutch, because there was no a yield really.

The fiber can die, yes, but it will yield control to some other fiber first and will invoke the trigger. It can't be reused/deleted without yields AFAIK. It might serve some next requests arrived in the same event loop iteration (see `fiber_pool_f()`), but they won't see the transaction, because `tx_end_msg()` clears it.

Reply:
> The fiber can die, yes, but it will yield control to some other fiber first and will invoke the trigger. It can't be reused/deleted without yields AFAIK.

This is specific to fiber pool implementation AFAIU (fiber_pool_f). Should we change it, e.g. change idle_timeout so that a fiber that doesn't have anything to do isn't reused but simply dies, it will not execute on_yield triggers and a detached transaction will not get aborted in non-mvcc case.

We need to find a way to detach a transaction from a fiber in such a way that neither of them has any reference to each other. Can't we look at txn->flags and abort a transaction on detach if it doesn't contain TXN_CAN_YIELD?

Reply:
Fixed